File: src/lib/today.ts
import { store } from "./store";
import { computeLedger, suggestCuts } from "@/core/ledger";
import { bestFit, findGaps } from "@/core/gaps";
import { questsForDay } from "@/core/game";
import { sharedGaps } from "@/core/overlap";
import { fmt, fromDate } from "@/core/time";
import { MODE_RULES } from "@/core/types";
import { buildJourney, BUILDINGS } from "./journey";
import { clock } from "@/services/prt";
import { transitNeed } from "@/core/transitRelevance";
import { REGION } from "@/services/schedule";

/** Who publishes the timetable we just quoted, per region. */
const AGENCY = {
  hudson: "NJ TRANSIT and the Port Authority of New York and New Jersey",
  oakland: "Pittsburgh Regional Transit",
} as const;

const isPlace = (p?: string): p is keyof typeof BUILDINGS => !!p && p in BUILDINGS;

/**
 * Everything the Today screen needs.
 *
 * The bus block comes from the same buildJourney() the map screen uses, so the
 * card and the map can never disagree about when to leave. Two code paths
 * computing the same time is how a demo ends up showing 14:02 on one screen
 * and 14:07 on the next.
 */
export async function buildToday(opts?: { to?: string }) {
  const s = store();
  const horizon = MODE_RULES[s.mode].horizonHours;
  const now = new Date();
  const liveTasks = s.tasks.filter((t) => !t.completedAt).filter((t) => {
    if (s.mode === "crisis") return t.domain === "learn" || t.domain === "build";
    if (s.mode === "chill") return t.dueAt ? (t.dueAt.getTime() - now.getTime()) / 36e5 <= (horizon ?? 48) : false;
    return true;
  });

  // Resolved before the windows are built, because a plan written at wake and
  // never moved will still be offering this morning's window at seven at night.
  const c = clock();
  const nowMin = Math.floor(c.sec / 60);

  const ledger = computeLedger(s.blocks, s.profile, s.travel, liveTasks, s.estimator);
  const gaps = findGaps(s.blocks, s.profile, s.travel, nowMin);
  const picks = new Map(gaps.map((g) => [g.id, bestFit(g, liveTasks, s.estimator)] as const));
  const seen = new Set<string>();
  for (const [id, task] of picks) {
    if (task && seen.has(task.id)) picks.set(id, undefined);
    if (task) seen.add(task.id);
  }

  // Which leg matters right now: getting to the next class, or getting home
  // after the last one.
  const ordered = [...s.blocks].sort((a, b) => a.start - b.start);
  const nextClass = ordered.find((b) => b.start > nowMin && isPlace(b.place));
  const lastClass = [...ordered].reverse().find((b) => isPlace(b.place));

  // The timeline the clients draw. Status and progress are decided here, never
  // on the device: clock() may be the simulated demo clock, so a phone asking
  // Date() mid-rehearsal would mark the 2:30 seminar finished at seven in the
  // evening. Additive field; the web page ignores it.
  const firstUpcoming = ordered.find((b) => b.start > nowMin);
  const timeline = ordered.map((b) => ({
    id: b.id,
    title: b.title,
    kind: b.kind,
    courseCode: b.courseCode ?? null,
    place: b.place ?? null,
    startText: fmt(b.start),
    endText: fmt(b.end),
    minutes: b.end - b.start,
    status:
      b.end <= nowMin ? "done"
      : b.start <= nowMin ? "now"
      : b.id === firstUpcoming?.id ? "next"
      : "later",
    progress: b.start <= nowMin && nowMin < b.end ? +((nowMin - b.start) / (b.end - b.start)).toFixed(3) : null,
    remainingMinutes: b.start <= nowMin && nowMin < b.end ? b.end - nowMin : null,
  }));

  // Demand-driven: a bus is worked out when there is a reason to catch one,
  // not on every load of Today. Idle means no protobuf feeds fetched, no route
  // solved, and no bus card on a screen where a bus is not the answer.
  const need = transitNeed({ nowMin, blocks: s.blocks, askedFor: opts?.to, isPlace });
  const goingToClass = need.reason === "class" || (need.reason === "asked" && need.to !== "Home");

  // A trip needs a destination. transitNeed always supplies one for "class"
  // and "asked", so the guard is belt-and-braces -- but the old fallback was
  // a hardcoded "Cathedral", which is a building in Pittsburgh, and a
  // Pittsburgh building in a New Jersey journey is not a default, it is a bug
  // waiting for the one code path that reaches it.
  const leg = !need.needed || (need.reason !== "home" && !need.to)
    ? undefined
    : need.reason === "home"
      ? { from: (need.block?.place ?? lastClass?.place) as keyof typeof BUILDINGS, to: "Home" as const, arriveBySec: undefined, why: "home after your last class" }
      : {
          from: "Home" as const,
          to: need.to as keyof typeof BUILDINGS,
          arriveBySec: need.block ? need.block.start * 60 : undefined,
          why: need.block ? `to ${need.block.title}` : `to ${need.to}`,
        };

  const journey = leg ? await buildJourney({ from: leg.from, to: leg.to, arriveBySec: leg.arriveBySec, now: c }) : undefined;
  const o = journey?.options[0];

  const busQuest = o && gaps.length ? { gapId: gaps[gaps.length - 1].id, leaveBy: Math.floor(o.leaveBySec / 60), route: o.route } : undefined;
  const quests = questsForDay(gaps, picks, busQuest);
  const shared = sharedGaps(gaps, s.friends);
  const cuts = ledger.overCommitted ? suggestCuts(liveTasks, -ledger.slack, new Set(s.profile.priorityDomains), s.estimator) : [];

  return {
    mode: s.mode,
    user: s.user,
    ledger,
    blocks: timeline,
    gaps: gaps.map((g) => ({ ...g, startText: fmt(g.start), endText: fmt(g.end), pick: picks.get(g.id) ?? null })),
    quests: quests.map((q) => ({ ...q, expiresText: fmt(q.expiresAt) })),
    bus: journey && o && leg
      ? {
          route: o.route,
          headsign: o.headsign,
          leaveBySec: o.leaveBySec,
          leaveByText: o.leaveByText,
          departsText: o.departsText,
          arrivalText: o.arriveText,
          status: o.status,
          live: o.status === "live",
          ghost: o.status === "ghost",
          delaySec: o.delaySec,
          from: leg.from,
          to: leg.to,
          why: leg.why,
          stopName: journey.boardStop.name,
          alightName: journey.alightStop.name,
          walkToStop: journey.walkToStop.minutes,
          walkToDest: journey.walkToDest.minutes,
          rideMinutes: o.rideMinutes,
          vehicleKm: o.vehicle ? +(o.vehicle.metersToStop / 1000).toFixed(1) : null,
          verdict: journey.destination.arriveByText ? o.verdict : null,
          classAtText: journey.destination.arriveByText ?? null,
          /** Deep link into the full map with this exact leg selected. */
          mapHref: `/map?from=${leg.from}&to=${leg.to}${journey.destination.arriveByText ? `&arriveBy=${journey.destination.arriveByText}` : ""}`,
        }
      : null,
    arrivals: (journey?.options ?? []).map((x) => ({
      route: x.route, text: x.departsText, realtime: x.status === "live", status: x.status, headsign: x.headsign,
      leaveByText: x.leaveByText, arriveText: x.arriveText,
      // Null verdict means there is nothing to be late for, which is not the
      // same as making it. Clients must be able to tell those apart.
      makesIt: x.verdict?.makesIt ?? null, marginMin: x.verdict?.marginMin ?? null,
      totalMinutes: x.totalMinutes, waitMinutes: x.waitMinutes,
    })),
    ghosts: (journey?.options ?? []).filter((x) => x.status === "ghost").map((x) => ({ route: x.route })),
    transit: {
      clockText: fmt(nowMin),
      simulated: c.simulated,
      realtimeOk: journey?.realtime.tripsOk ?? false,
      walkSource: journey?.walkToStop.source ?? "estimate",
      // Why there is or is not a bus, so a client shows a sentence rather than
      // an empty card, and nobody has to guess whether it is broken or idle.
      reason: need.reason,
      why: need.why,
      planned: need.needed,
      // Named here, where the region is known, so nothing downstream has to
      // guess. The factsheet used to hardcode "Pittsburgh Regional Transit"
      // into the source of every bus fact, which meant Orbit defended a
      // Hoboken light rail departure by citing an agency in another state --
      // in the one feature built to prove it is not bluffing.
      agency: AGENCY[REGION],
      alerts: journey?.alerts ?? [],
      alertsOk: journey?.realtime.alertsOk ?? false,
    },
    shared: shared.map((w) => ({ ...w, startText: fmt(w.start), endText: fmt(w.end), names: w.userIds.map((id) => s.friends.find((f) => f.userId === id)?.name ?? id) })),
    tasks: liveTasks.map((t) => ({ ...t, planningMinutes: s.estimator.planningMinutes(t) })),
    cuts,
    calibration: s.estimator.calibration(),
    proposals: s.proposals,
    events: s.events.slice(-30).reverse(),
  };
}

export { fromDate };

File: ios/Orbit/Models/Today.swift
import Foundation

/// Mirrors `GET /api/today` (`src/lib/today.ts`). Every number here was
/// computed on the server. Nothing in this file does arithmetic on a time, and
/// nothing downstream of it should either: if a view needs a new number, it is
/// added to `buildToday()` and to `docs`/`DECISIONS.md`, not derived here.
///
/// Only the fields the iOS client renders are declared. Unknown JSON keys are
/// ignored by `JSONDecoder`, so the server staying ahead of the app is safe;
/// the app inventing a field it wants is not.
struct Today: Decodable {
    let mode: Mode
    let user: User
    let ledger: Ledger
    /// Added for the iOS timeline. Optional so a build of the app still runs
    /// against a server deployed before the field existed — it degrades to the
    /// gaps-only layout instead of showing an empty day.
    let blocks: [DayBlock]?
    let gaps: [Gap]
    let quests: [Quest]
    let bus: Bus?
    let shared: [SharedWindow]
    let tasks: [TaskItem]
    let transit: Transit
    let proposals: [Proposal]
    /// What the weekly review has worked out about this student. Optional so a
    /// build still decodes against a server deployed before the field existed;
    /// empty renders nothing rather than an "all clear" card.
    /// What to let go of when the day will not fit. Empty unless
    /// `ledger.overCommitted`.
    let cuts: [Cut]?
    let learned: [LearnedFact]?
    /// Deadlines the remaining work no longer fits in front of.
    let atRisk: [AtRiskTask]?

    enum Mode: String, Decodable, CaseIterable {
        case normal, crisis, chill
        var label: String {
            switch self {
            case .normal: return "Normal"
            case .crisis: return "Crisis"
            case .chill: return "Chill"
            }
        }
    }

    struct User: Decodable {
        let name: String
        let xpWeek: Int
        let streakWeeks: Int
        let group: String?
    }

    struct Ledger: Decodable {
        let usable: Int
        let naiveFree: Int
        /// Minutes between waking and sleeping. `fixed + travel + meals +
        /// routines + usable` sums to exactly this, which is what makes it the
        /// right whole for the ledger rings. `naiveFree` is NOT that whole —
        /// it already has class taken out. Optional so a build still decodes
        /// against a server deployed before the field existed.
        let awake: Int?
        let travel: Int
        let meals: Int
        let routines: Int
        /// travel + meals + routines, summed by computeLedger(). Optional only so
        /// an older deploy still decodes; the fallback sums the same three.
        let friction: Int?
        let fixed: Int
        let queued: Int
        let slack: Int
        let overCommitted: Bool

        var frictionMinutes: Int { friction ?? (travel + meals + routines) }
    }

    struct DayBlock: Decodable, Identifiable, Hashable {
        let id: String
        let title: String
        let kind: String            // class | work | event | routine
        let courseCode: String?
        let place: String?
        let startText: String
        let endText: String
        let minutes: Int
        let status: Status
        /// 0...1 while in progress, nil otherwise. Server-computed.
        let progress: Double?
        let remainingMinutes: Int?

        /// Forward-compatible: a status this build has never heard of sorts in
        /// with the rest of the day rather than crashing the whole screen.
        enum Status: String, Decodable, Hashable {
            case now, next, done, later

            init(from decoder: Decoder) throws {
                let raw = try decoder.singleValueContainer().decode(String.self)
                self = Status(rawValue: raw) ?? .later
            }
        }
    }

    struct Gap: Decodable, Identifiable, Hashable {
        let id: String
        let startText: String
        let endText: String
        let usable: Int
        let fromPlace: String?
        let isEvening: Bool
        let pick: Pick?

        struct Pick: Decodable, Hashable {
            let id: String
            let title: String
            let estimateMinutes: Int
        }
    }

    struct Quest: Decodable, Identifiable, Hashable {
        let id: String
        let title: String
        let xp: Int
        let kind: String
        let expiresText: String
    }

    struct Bus: Decodable, Hashable {
        let route: String
        let headsign: String
        let leaveByText: String
        let arrivalText: String
        let status: String          // live | scheduled | ghost
        let live: Bool
        let ghost: Bool
        let why: String
        let stopName: String
        let walkToStop: Int
        let rideMinutes: Int
        let classAtText: String?
        let verdict: Verdict?

        struct Verdict: Decodable, Hashable {
            let makesIt: Bool
            let marginMin: Int
        }
    }

    struct SharedWindow: Decodable, Hashable {
        let startText: String
        let endText: String
        let minutes: Int
        let names: [String]
    }

    struct TaskItem: Decodable, Identifiable, Hashable {
        let id: String
        let title: String
        let domain: String
        let estimateMinutes: Int
        let planningMinutes: Int
        let courseCode: String?
        let dueAt: String?
    }

    struct Transit: Decodable, Hashable {
        let clockText: String
        let simulated: Bool
        let realtimeOk: Bool
        /// Why there is nothing to catch. The server writes the sentence; the
        /// app prints it. Present whenever `bus` is nil, which is most of the
        /// day, and the reason "Getting there" is never an empty card.
        let why: String?
        /// The class a departure would be for, when one is still ahead.
        let nextClass: NextClass?

        struct NextClass: Decodable, Hashable {
            let title: String
            let startText: String
            let minutesAway: Int
        }
    }

    struct Cut: Decodable, Identifiable, Hashable {
        let task: CutTask
        let minutesSaved: Int

        struct CutTask: Decodable, Hashable {
            let title: String
        }

        var id: String { task.title }
    }

    struct LearnedFact: Decodable, Identifiable, Hashable {
        let aspect: String
        let label: String
        /// Written by code on the server precisely so the words can never
        /// disagree with the multiplier behind them. The app prints it and
        /// never rephrases it. See DECISIONS, 20 Sept 03:05.
        let sentence: String

        /// One aspect can speak about several categories, so the sentence is
        /// part of the identity.
        var id: String { aspect + "::" + sentence }
    }

    struct AtRiskTask: Decodable, Identifiable, Hashable {
        let taskId: String
        let title: String
        let needsMinutes: Int

        var id: String { taskId }
    }

    struct Proposal: Decodable, Identifiable, Hashable {
        let id: String
        let status: String
        let proposal: Body

        struct Body: Decodable, Hashable {
            let kind: String
            let reason: String
            let body: String?
            let to: String?
            let subject: String?
            let building: String?
            let message: String?
        }
    }

    var pendingProposals: [Proposal] { proposals.filter { $0.status == "pending" } }
}

// MARK: - Other endpoints

struct PlanResponse: Decodable {
    let proposals: [Today.Proposal]
    let narration: String
    let provider: String
}

struct DecisionResponse: Decodable {
    let ok: Bool
    let status: String?
    let effect: String?
}

/// `POST /api/complete`. XP and its reasons are the server's answer, never the
/// phone's guess.
struct CompleteResponse: Decodable {
    let ok: Bool
    let xp: Int
    let reasons: [String]
    let planned: Int
    let multiplier: Double
}

struct ModeResponse: Decodable {
    let ok: Bool
    let mode: Today.Mode?
}
